// xslt-processor/dom/functions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.xmlParse = exports.domCreateDTDSection = exports.domCreateDocumentFragment = exports.domCreateComment = exports.domCreateCDATASection = exports.domCreateElement = exports.domCreateTransformedTextNode = exports.domCreateTextNode = exports.domAppendTransformedChild = exports.domAppendChild = exports.domSetTransformedAttribute = exports.domSetAttribute = exports.domGetAttributeValue = void 0;
// Copyright 2023 Design Liquido
// Copyright 2018 Johannes Wilm
// Copyright 2005 Google Inc.
// All Rights Reserved
//
// Original author: Steffen Meschkat <mesch@google.com>
//
// An XML parse and a minimal DOM implementation that just supports
// the subset of the W3C DOM that is used in the XSLT implementation.
var he_1 = __importDefault(require("he"));
var util_1 = require("./util");
var xdocument_1 = require("./xdocument");
var xmltoken_1 = require("./xmltoken");
var XML10_TAGNAME_REGEXP = new RegExp("^(".concat(xmltoken_1.XML10_NAME, ")"));
var XML10_ATTRIBUTE_REGEXP = new RegExp(xmltoken_1.XML10_ATTRIBUTE, 'g');
var XML11_TAGNAME_REGEXP = new RegExp("^(".concat(xmltoken_1.XML11_NAME, ")"));
var XML11_ATTRIBUTE_REGEXP = new RegExp(xmltoken_1.XML11_ATTRIBUTE, 'g');
// Wrapper around DOM methods so we can condense their invocations.
function domGetAttributeValue(node, name) {
    return node.getAttributeValue(name);
}
exports.domGetAttributeValue = domGetAttributeValue;
function domSetAttribute(node, name, value) {
    return node.setAttribute(name, value);
}
exports.domSetAttribute = domSetAttribute;
function domSetTransformedAttribute(node, name, value) {
    return node.setTransformedAttribute(name, value);
}
exports.domSetTransformedAttribute = domSetTransformedAttribute;
function domAppendChild(node, child) {
    return node.appendChild(child);
}
exports.domAppendChild = domAppendChild;
function domAppendTransformedChild(node, child) {
    return node.appendTransformedChild(child);
}
exports.domAppendTransformedChild = domAppendTransformedChild;
function domCreateTextNode(node, text) {
    return node.createTextNode(text);
}
exports.domCreateTextNode = domCreateTextNode;
function domCreateTransformedTextNode(node, text) {
    return node.createTransformedTextNode(text);
}
exports.domCreateTransformedTextNode = domCreateTransformedTextNode;
function domCreateElement(doc, name) {
    return doc.createElement(name);
}
exports.domCreateElement = domCreateElement;
function domCreateCDATASection(doc, data) {
    return doc.createCDATASection(data);
}
exports.domCreateCDATASection = domCreateCDATASection;
function domCreateComment(doc, text) {
    return doc.createComment(text);
}
exports.domCreateComment = domCreateComment;
function domCreateDocumentFragment(doc) {
    return doc.createDocumentFragment();
}
exports.domCreateDocumentFragment = domCreateDocumentFragment;
function domCreateDTDSection(doc, data) {
    return doc.createDTDSection(data);
}
exports.domCreateDTDSection = domCreateDTDSection;
/**
 * Parses the given XML string with our custom, JavaScript XML parser
 * @param xml The XML String.
 * @returns A XDocument.
 * @author Steffen Meschkat <mesch@google.com>
 */
function xmlParse(xml) {
    var regexEmpty = /\/$/;
    var regexTagname;
    var regexAttribute;
    if (xml.match(/^<\?xml/)) {
        // When an XML document begins with an XML declaration
        // VersionInfo must appear.
        if (xml.search(new RegExp(xmltoken_1.XML10_VERSION_INFO)) == 5) {
            regexTagname = XML10_TAGNAME_REGEXP;
            regexAttribute = XML10_ATTRIBUTE_REGEXP;
        }
        else if (xml.search(new RegExp(xmltoken_1.XML11_VERSION_INFO)) == 5) {
            regexTagname = XML11_TAGNAME_REGEXP;
            regexAttribute = XML11_ATTRIBUTE_REGEXP;
        }
        else {
            // VersionInfo is missing, or unknown version number.
            // TODO : Fallback to XML 1.0 or XML 1.1, or just return null?
            throw new Error('VersionInfo is missing, or unknown version number.');
        }
    }
    else {
        // When an XML declaration is missing it's an XML 1.0 document.
        regexTagname = XML10_TAGNAME_REGEXP;
        regexAttribute = XML10_ATTRIBUTE_REGEXP;
    }
    var xmldoc = new xdocument_1.XDocument();
    var root = xmldoc;
    var stack = [];
    var parent = root;
    stack.push(parent);
    var tag = false, quotes = false, doublequotes = false, start = 0;
    for (var i = 0; i < xml.length; ++i) {
        var char = xml.charAt(i);
        if (tag && !doublequotes && char === "'") {
            quotes = !quotes;
        }
        else if (tag && !quotes && char === '"') {
            doublequotes = !doublequotes;
        }
        else if (tag && char === '>' && !quotes && !doublequotes) {
            var text = xml.slice(start, i);
            if (text.charAt(0) == '/') {
                stack.pop();
                parent = stack[stack.length - 1];
            }
            else if (text.charAt(0) === '?') {
                // Ignore XML declaration and processing instructions
            }
            else if (text.charAt(0) === '!') {
                // Ignore comments
                // console.log(`Ignored ${text}`);
            }
            else {
                var empty = text.match(regexEmpty);
                var tagname = regexTagname.exec(text)[1];
                var node = domCreateElement(xmldoc, tagname);
                var attribute = void 0;
                while ((attribute = regexAttribute.exec(text))) {
                    var val = he_1.default.decode(attribute[5] || attribute[7] || '');
                    domSetAttribute(node, attribute[1], val);
                }
                node.siblingPosition = parent.childNodes.length;
                domAppendChild(parent, node);
                if (!empty) {
                    parent = node;
                    stack.push(node);
                }
                var namespaceMap = (0, util_1.namespaceMapAt)(node);
                if (node.prefix !== null) {
                    if (node.prefix in namespaceMap)
                        node.namespaceUri = namespaceMap[node.prefix];
                    // else, prefix is undefined. do anything?
                }
                else {
                    if ('' in namespaceMap)
                        node.namespaceUri = namespaceMap[''];
                }
                for (var i_1 = 0; i_1 < node.attributes.length; ++i_1) {
                    if (node.attributes[i_1].prefix !== null) {
                        if (node.attributes[i_1].prefix in namespaceMap) {
                            node.attributes[i_1].namespaceUri = namespaceMap[node.attributes[i_1].prefix];
                        }
                        // else, prefix undefined.
                    }
                    // elements with no prefix always have no namespace, so do nothing here.
                }
            }
            start = i + 1;
            tag = false;
            quotes = false;
            doublequotes = false;
        }
        else if (!tag && char === '<') {
            var text = xml.slice(start, i);
            if (text && parent !== root) {
                domAppendChild(parent, domCreateTextNode(xmldoc, text));
            }
            if (xml.slice(i + 1, i + 4) === '!--') {
                var endTagIndex = xml.slice(i + 4).indexOf('-->');
                if (endTagIndex) {
                    var node = domCreateComment(xmldoc, xml.slice(i + 4, i + endTagIndex + 4));
                    domAppendChild(parent, node);
                    i += endTagIndex + 6;
                }
            }
            else if (xml.slice(i + 1, i + 9) === '![CDATA[') {
                var endTagIndex = xml.slice(i + 9).indexOf(']]>');
                if (endTagIndex) {
                    var node = domCreateCDATASection(xmldoc, xml.slice(i + 9, i + endTagIndex + 9));
                    domAppendChild(parent, node);
                    i += endTagIndex + 11;
                }
            }
            else if (xml.slice(i + 1, i + 9) === '!DOCTYPE') {
                var endTagIndex = xml.slice(i + 9).indexOf('>');
                if (endTagIndex) {
                    var dtdValue = xml.slice(i + 9, i + endTagIndex + 9).trimStart();
                    // TODO: Not sure if this is a good solution.
                    // Trying to implement this: https://github.com/DesignLiquido/xslt-processor/issues/30
                    var node = void 0;
                    if (parent.nodeName === 'xsl:text') {
                        node = domCreateTextNode(xmldoc, "<!DOCTYPE ".concat(dtdValue, ">"));
                    }
                    else {
                        node = domCreateDTDSection(xmldoc, dtdValue);
                    }
                    domAppendChild(parent, node);
                    i += endTagIndex + dtdValue.length + 5;
                }
            }
            else {
                tag = true;
            }
            start = i + 1;
        }
    }
    return root;
}
exports.xmlParse = xmlParse;
//XDocument.prototype = new XNode(DOM_DOCUMENT_NODE, '#document');
